Add batch removal for remove par

Wegatriespython opened the pull request:
This PR uses a batch removal method for remove_par 

Linked to [PR 403](https://github.com/iiasa/ixmp_source/pull/403) in ixmp_source, this PR adds a batch method to improve performance on remove_par. 

For longer context, the current implementation of remove_par in ixmp/backend/jdbc.py via item_delete_elements() in the JDBC class, iterates through each K,V pair and calls the backend removeElement for each element. This results in a O(n^2) operation as the backend scans through the element list for each removal O(n), which when called n times results in an O(n^2) operation.

To remedy this performance bottleneck a new batch removal method was added to ixmp_source in the aforementioned PR 403, now in jdbc.py when the length of removals is >1 we call the batch removal method instead of the single item removal. 

Attached in this PR is the compiled jar file.  


## How to review

To review this PR, read the code diff, check the PR on ixmp_source. Verify all tests pass and run through actual scenarios. Recommendation is to allow for this branch to be opt-in to collect real use-case data to see if there are any un-intended side-effects from the java recompilation before merging into main. 



## PR checklist

<!-- This item is always required. -->
- [x] Continuous integration checks all ✅
<!--
The following items are *required* if the PR results in changes to user-facing
behaviour—such as new features, or fixes to existing behaviour.

They are *optional* if the changes are solely to documentation, test/CI
configuration, etc. In such cases, strike them out and add a short explanation,
for example:

- ~Add or expand tests.~ No change in behaviour, simply refactoring.
-->
- [x] Add or expand tests; coverage checks both ✅
- [x] Add, expand, or update documentation.
- [x] Update release notes.
  <!--
  To do this, add a single line at the TOP of the “Next release” section of
  RELEASE_NOTES.rst, where '999' is the GitHub pull request number:

  - Title or single-sentence description from above (:pull:`999`:).

  Commit with a message like “Add #999 to release notes”
  -->

Wegatriespython committed:
Fix slow remove par

Wegatriespython committed:
Bump jar file to latest build

Wegatriespython committed:
Fix line length

Wegatriespython committed:
Fix test failures

Wegatriespython committed:
Fix excess memory use

- Batch calls can have unbounded memory use on large removals.
- Add a batch size of 10,000 to keep memory usage bounded.

Wegatriespython committed:
Bump batch size to 1e6

glatterf42 reviewed (commented):
Overall, it will be fantastic once we get this to work :)
I think we can slightly improve on the changes still, see my comments. And I'm wondering whether we need to do anything else for this PR to be coupled to the `ixmp_source` PR. Let me check this step that @khaeru suggested:

> Look for the most recent ixmp PR/commit(s) that updated the ixmp.jar file. Figure out how they were created. I think https://github.com/iiasa/ixmp_source/blob/master/build_and_commit_jar.sh was used, but am not sure. I am not sure if this script still works. Possibly the steps could be done manually, since we will not be doing this often.

Please note that #604 will resolve the various issues in the Windows CI tests, I would not expect this PR to do so :)

glatterf42 commented on ixmp/backend/jdbc.py:
```diff
@@ -1341,8 +1342,27 @@ def item_delete_elements(
+        while True:
+            batch = list(islice(keys_iter, BATCH_SIZE))
+            if not batch:
+                break
```
(Note for future reference: `itertools` got [`batched()`](https://docs.python.org/3/library/itertools.html#itertools.batched) with Python 3.12, so once this becomes our oldest supported version, we can formally switch to that instead of `islice()`.)

Wegatriespython committed:
Apply micro-optimizations

- For primitive data types np.array can directly copy without
elementwise iteration
- For strings iteration required since java strings are objects. Faster
to directly pass to pd.series than to use [:] and rely on jpype to
convert

Wegatriespython commented on ixmp/backend/jdbc.py:
```diff
@@ -1341,8 +1342,27 @@ def item_delete_elements(
+        BATCH_SIZE = 10000
+
+        # Use islice for memory-efficient iteration without materializing all keys
+        ArrayList = jpype.JClass("java.util.ArrayList")
```
Arraylist is faster, I tested it and this source also backs it up [https://stackoverflow.com/questions/322715/when-to-use-linkedlist-over-arraylist-in-java#:~:text=Summary%20ArrayList%20with,different%20algorithmic%20runtimes.](url)

Wegatriespython commented on ixmp/backend/jdbc.py:
```diff
@@ -1341,8 +1342,27 @@ def item_delete_elements(
+                key_vectors = ArrayList()
+                for key in batch:
+                    key_vectors.add(to_jlist(key))
+                jitem.removeElements(key_vectors)
```
It produces a type error, because  JPype cannot convert nested Python lists (list of lists) directly to Java collections via to_jlist(). The explicit loop is necessary to convert each individual key to a Java LinkedList before adding it to the collection.

`       File "/home/raghunathan/ixmp/test_direct_batch.py", line 47, in <module>
         result = to_jlist(batch)
                  ^^^^^^^^^^^^^^^
       File "/home/raghunathan/ixmp/ixmp/backend/jdbc.py", line 1594, in to_jlist
         return java.LinkedList(arg)
                ^^^^^^^^^^^^^^^^^^^^
       File "/home/raghunathan/ixmp/.venv/lib/python3.12/site-packages/jpype/protocol.py", line 79, in _JSequenceConvert
         return _jclass.JClass('java.util.Arrays').asList(obj)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     TypeError: No matching overloads found for *static* java.util.Arrays.asList(list), options are:
        public static java.util.List java.util.Arrays.asList(java.lang.Object[])`

meksor replied:
To improve performance here you could probably `ArrayList.ensureCapacity()` in advance to avoid memory allocations within the loop.

glatterf42 replied:
I guess another approach would be to adapt the `to_jlist()` function, because this is [part of our package here](https://github.com/iiasa/ixmp/blob/a8a5eb6853971ead86ad4862e268bb832dbd859e/ixmp/backend/jdbc.py#L1535), so we can make it do what is useful for us -- maybe even handle nested lists correctly. But if you don't think the performance gain here is worth it, no need to spend too much time.

Wegatriespython committed:
Address PR comments

Wegatriespython committed:
Fix linelength

Wegatriespython committed:
Bump jar file

glatterf42 commented:
> [![CLA assistant check](https://camo.githubusercontent.com/bb14f9741a98be03eb447ca21a4a5e96201df401e5be70f9ef52130f4e1ff903/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/iiasa/ixmp?pullRequest=603) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/iiasa/ixmp?pullRequest=603) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/iiasa/ixmp?pullRequest=603) it.

Thanks for the update :)
Good to know this is getting into shape :)
Before we can merge this, please do sign the CLA as requested above :)

Wegatriespython committed:
Ruff formatting

glatterf42 commented:
If by "this script" you're referring to https://github.com/iiasa/ixmp_source/blob/master/build_and_commit_jar.sh, please see [my assessment above](https://github.com/iiasa/ixmp/pull/603#issuecomment-3422123442) that I don't think it will be necessary to run this exact script (and therefore neither to install docker).

If you're talking about another script, please let me know which one :)

glatterf42 committed:
DRY changes and restore faster JPype path

Wegatriespython committed:
Bump jar file

Source diff https://github.com/iiasa/ixmp_source/compare/2a7433a...fef6b59

Wegatriespython committed:
Add #603 to release notes

Wegatriespython committed:
Refresh downscoped batch removal jar

khaeru reviewed (approved):
LGTM!
